src/jit/codegenxarch.cpp

```diff
@@ -4728,6 +4754,47 @@ void CodeGen::genCodeForShift(GenTreePtr tree)
     regNumber  operandReg = operand->gtRegNum;
 
     GenTreePtr shiftBy = tree->gtGetOp2();
+#ifndef _TARGET_64BIT_
+    if (operand->OperGet() == GT_LONG)
+    {
+        GenTree* operandLo = operand->gtGetOp1();
+        GenTree* operandHi = operand->gtGetOp2();
+
+        regNumber regLo = operandLo->gtRegNum;
+        regNumber regHi = operandHi->gtRegNum;
+
+        assert(shiftBy->IsCnsIntOrI());
+
+        int count = (int)shiftBy->AsIntConCommon()->IconValue();
+
+        assert(tree->OperGet() == GT_LSH_HI || tree->OperGet() == GT_RSH_LO);
+
+        if (tree->OperGet() == GT_LSH_HI)
+        {
+            inst_RV_RV_IV(INS_shld, EA_4BYTE, regHi, regLo, count);
+
+            // We have to do this after the shift, because the lo reg may be the
+            // same as the tree's target reg, and we don't want to overwrite it.
+            if (regHi != tree->gtRegNum)
+            {
+                inst_RV_RV(INS_mov, tree->gtRegNum, regHi, targetType);
+            }
+        }
+        else
+        {
+            assert(tree->OperGet() == GT_RSH_LO);
+            inst_RV_RV_IV(INS_shrd, EA_4BYTE, regLo, regHi, count);
+
+            // We have to do this after the shift, because the hi reg may be the
+            // same as the tree's target reg, and we don't want to overwrite it.
+            if (regLo != tree->gtRegNum)
+            {
+                inst_RV_RV(INS_mov, tree->gtRegNum, regLo, targetType);
+            }
+        }
```

Comment:
Might be useful to tail merge this and the above by setting a regResult in each case and then handling the 'mov' just once based on regResult.